packages/@aws-cdk/aws-ec2/lib/cfn-init-elements.ts

```diff
@@ -0,0 +1,405 @@
+   */
+  FILE = 'FILE',
+
+  /**
+   * A command
+   */
+  COMMAND = 'COMMAND',
+
+  /**
+   * A service
+   */
+  SERVICE = 'SERVICE',
+}
+
+/**
+ * An object that represents reasons to restart an InitService
+ *
+ * Pass an instance of this object to all the `InitFile`, `InitCommand`,
+ * `InitSource` and `InitPackage` objects that you want to restart
+ * a service, and finally to the `InitService` itself as well.
+ */
+export class InitServiceRestartHandle {
+  private readonly commands = new Array<string>();
+  private readonly files = new Array<string>();
+  private readonly sources = new Array<string>();
+  private readonly packages: Record<string, string[]> = {};
+
+  /**
+   * Add a command key to the restart set
+   */
+  public addCommand(key: string) {
+    return this.commands.push(key);
+  }
+
+  /**
+   * Add a file key to the restart set
+   */
+  public addFile(key: string) {
+    return this.files.push(key);
+  }
+
+  /**
+   * Add a source key to the restart set
+   */
+  public addSource(key: string) {
+    return this.sources.push(key);
+  }
+
+  /**
+   * Add a package key to the restart set
+   */
+  public addPackage(packageType: string, key: string) {
+    if (!this.packages[packageType]) {
+      this.packages[packageType] = [];
+    }
+    this.packages[packageType].push(key);
+  }
+
+  /**
+   * Render the restart handles for use in an InitService declaration
+   */
+  public renderRestartHandles(): any {
+    const nonEmpty = <A>(x: A[]) => x.length > 0 ? x : undefined;
+
+    return {
+      commands: nonEmpty(this.commands),
+      files: nonEmpty(this.files),
+      packages: Object.keys(this.packages).length > 0 ? this.packages : undefined,
+      sources: nonEmpty(this.sources),
+    };
+  }
+}
+
+/**
+ * Context information passed when an InitElement is being consumed
+ */
+export interface InitBindOptions {
+  /**
+   * Scope in which to define any resources, if necessary.
+   */
+  readonly scope: Construct;
+
+  /**
+   * Which OS platform (Linux, Windows) the init block will be for.
+   * Impacts which config types are available and how they are created.
+   */
+  readonly platform: InitPlatform;
+
+  /**
+   * Ordered index of current element type. Primarily used to auto-generate
+   * command keys and retain ordering.
+   */
+  readonly index: number;
+
+  /**
+   * Instance role of the consuming instance or fleet
+   */
+  readonly instanceRole: iam.IRole;
+}
+
+/**
+ * A return type for a configured InitElement. Both its CloudFormation representation, and any
+ * additional metadata needed to create the CloudFormation:Init.
+ */
+export interface InitElementConfig {
+  /**
+   * The CloudFormation representation of the configuration of an InitElement.
+   */
+  readonly config: Record<string, any>;
+
+  /**
+   * Optional authentication blocks to be associated with the Init Config
+   *
+   * @default - No authentication associated with the config
+   */
+  readonly authentication?: Record<string, any>;
```

Comment:
CloudFormation details are leaking out of the L2. I would expect this to be abstracted away at this point.

Ideally, the result of the bind returns types that are codified at this level and converted into CloudFormation by the consumer (of the bind). However, I can't think of a way to do this without adding a bunch of additional types here.

Throwing this out there to consider.

Reply:
Another option here (thanks to @eladb) which I think fits this use case better is to mark the `bind()` methods as internal and not export the `InitElementBindOptions` and `InitElementConfig` interfaces, and then keep the current implementation. This would fix the leak.

The limitation is that a customer cannot extend and implement their own `InitElement` which I think is not a use case here.